PxMatrix.cpp

```diff
@@ -0,0 +1,955 @@
+    spi_frequency = PxMATRIX_MAX_SPI_FREQUENCY;
+  }
+  SPI.setFrequency(spi_frequency);
+#endif
+}
+
+uint32_t PxMATRIX::getPixelCount() { return _width * _height; }
+
+void PxMATRIX::spi_init() {
+#ifdef ESP32
+  SPI.begin(_SPI_CLK, _SPI_MISO, _SPI_MOSI, _SPI_SS);
+#else
+  SPI.begin();
+#endif
+
+#if defined(ESP32) || defined(ESP8266)
+  SPI.setFrequency(PxMATRIX_MIN_SPI_FREQUENCY);
+#endif
+
+  SPI.setDataMode(SPI_MODE0);
+  SPI.setBitOrder(MSBFIRST);
+}
+
+void PxMATRIX::begin(uint8_t row_pattern) {
+  _row_pattern = row_pattern;
+  if (_row_pattern == 4) _scan_pattern = ZIGZAG;
+  _mux_pattern = BINARY;
+  _color_order = RRGGBB;
+
+  _pattern_color_bytes = (_height / _row_pattern) * (_width / 8);
+  _row_sets_per_buffer = _rows_per_buffer / _row_pattern;
+  _send_buffer_size = _pattern_color_bytes * 3;
+
+  spi_init();
+
+  pinMode(_OE_PIN, OUTPUT);
+  pinMode(_LATCH_PIN, OUTPUT);
+  pinMode(_A_PIN, OUTPUT);
+  pinMode(_B_PIN, OUTPUT);
+  digitalWrite(_A_PIN, LOW);
+  digitalWrite(_B_PIN, LOW);
+  digitalWrite(_OE_PIN, HIGH);
+
+  if (_row_pattern >= 8) {
+    pinMode(_C_PIN, OUTPUT);
+    digitalWrite(_C_PIN, LOW);
+  }
+  if (_row_pattern >= 16) {
+    pinMode(_D_PIN, OUTPUT);
+    digitalWrite(_D_PIN, LOW);
+  }
+  if (_row_pattern >= 32) {
+    pinMode(_E_PIN, OUTPUT);
+    digitalWrite(_E_PIN, LOW);
+  }
+
+  // Precompute row offset values
+  for (uint8_t yy = 0; yy < _height; yy++)
+    _row_offset[yy] = ((yy) % _row_pattern) * _send_buffer_size + _send_buffer_size - 1;
+}
+
+void PxMATRIX::set_mux(uint8_t value) {
+  if (_mux_pattern == BINARY) {
+    if (value & 0x01)
+      digitalWrite(_A_PIN, HIGH);
+    else
+      digitalWrite(_A_PIN, LOW);
+    if (_mux_delay_A) delayMicroseconds(_mux_delay_A);
+
+    if (value & 0x02)
+      digitalWrite(_B_PIN, HIGH);
+    else
+      digitalWrite(_B_PIN, LOW);
+    if (_mux_delay_B) delayMicroseconds(_mux_delay_B);
+
+    if (_row_pattern >= 8) {
+      if (value & 0x04)
+        digitalWrite(_C_PIN, HIGH);
+      else
+        digitalWrite(_C_PIN, LOW);
+      if (_mux_delay_C) delayMicroseconds(_mux_delay_C);
+    }
+
+    if (_row_pattern >= 16) {
+      if (value & 0x08)
+        digitalWrite(_D_PIN, HIGH);
+      else
+        digitalWrite(_D_PIN, LOW);
+      if (_mux_delay_D) delayMicroseconds(_mux_delay_D);
+    }
+
+    if (_row_pattern >= 32) {
+      if (value & 0x10)
+        digitalWrite(_E_PIN, HIGH);
+      else
+        digitalWrite(_E_PIN, LOW);
+      if (_mux_delay_E) delayMicroseconds(_mux_delay_E);
+    }
+  }
+
+  if (_mux_pattern == STRAIGHT) {
+    if (value == 0)
+      digitalWrite(_A_PIN, LOW);
+    else
+      digitalWrite(_A_PIN, HIGH);
+
+    if (value == 1)
+      digitalWrite(_B_PIN, LOW);
+    else
+      digitalWrite(_B_PIN, HIGH);
+
+    if (value == 2)
+      digitalWrite(_C_PIN, LOW);
+    else
+      digitalWrite(_C_PIN, HIGH);
+
+    if (value == 3)
+      digitalWrite(_D_PIN, LOW);
+    else
+      digitalWrite(_D_PIN, HIGH);
+  }
```

Comment:
Same here, seems like the author is allergic to else if ladders, i'll stop pointing them out